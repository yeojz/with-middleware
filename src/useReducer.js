import { BATCH_ACTION, SET_ACTION } from './constants';

function batchAction(reducer, store, next, action) {
  const state = action.payload.reduce(
    (nextState, current) => reducer(nextState, current),
    action.state || store.getState()
  );

  return next(state);
}

function useReducer(reducer) {
  if (typeof reducer !== 'function') {
    throw new Error(
      `Expecting argument 1 of "useReducer" to be a function. ${typeof reducer} given.`
    );
  }

  return store => next => action => {
    if (action.type === BATCH_ACTION) {
      return batchAction(reducer, store, next, action);
    }

    if (action.type === SET_ACTION) {
      return next(action.payload);
    }

    const state = reducer(store.getState(), action);
    return next(state);
  };
}

export default useReducer;
