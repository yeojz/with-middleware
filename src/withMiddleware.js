import { createFactory } from 'react';
import applyMiddleware from './helpers/applyMiddleware';

function createComponent(stateName, dispatchName, createDispatch, factory) {
  function ComponentWithMiddleware(props) {
    const dispatch = createDispatch(props[stateName], props[dispatchName]);

    return factory(
      Object.assign({}, props, {
        [dispatchName]: dispatch
      })
    );
  }

  return ComponentWithMiddleware;
}

function withMiddleware(stateName, dispatchName, middlewares) {
  if (!Array.isArray(middlewares)) {
    throw new Error(
      `Expecting argument 3 of "withMiddleware" to be an array. ${typeof middlewares} given.`
    );
  }

  const createDispatch = applyMiddleware(...middlewares);

  return BaseComponent => {
    const factory = createFactory(BaseComponent);
    return createComponent(stateName, dispatchName, createDispatch, factory);
  };
}

export default withMiddleware;
