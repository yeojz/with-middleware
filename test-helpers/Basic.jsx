/* eslint-disable react/prop-types */

import React from 'react';

function Basic(props) {
  return (
    <div>
      <div data-testid="value">{props.counter}</div>
      <button
        data-testid="increment"
        onClick={() => props.setCounter(n => n + 1)}
      >
        Increment
      </button>
      <button
        data-testid="decrement"
        onClick={() => props.setCounter(n => n - 1)}
      >
        Decrement
      </button>
    </div>
  );
}

export default Basic;
