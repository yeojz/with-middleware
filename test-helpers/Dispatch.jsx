/* eslint-disable react/prop-types */
import React from 'react';

function Dispatch(props) {
  return (
    <div>
      <div data-testid="value">{props.state.count}</div>
      <button
        data-testid="increment"
        onClick={() => props.dispatch({ type: 'INCREMENT' })}
      >
        Increment
      </button>
      <button
        data-testid="decrement"
        onClick={() => props.dispatch({ type: 'DECREMENT' })}
      >
        Decrement
      </button>
    </div>
  );
}

export default Dispatch;
