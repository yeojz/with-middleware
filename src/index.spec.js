import React from 'react';
import { compose, withState } from 'recompose';
import { render, cleanup, fireEvent } from 'react-testing-library';
import useReducer from './useReducer';
import withMiddleware from './withMiddleware';
import Basic from '../tests/helpers/Basic';
import Dispatch from '../tests/helpers/Dispatch';
import reducer from '../tests/helpers/reducer';

function createRenderer(BaseComponent) {
  return (stateName, stateUpdaterName, initialState, middlewares = []) => {
    const Component = compose(
      withState(stateName, stateUpdaterName, initialState),
      withMiddleware(stateName, stateUpdaterName, middlewares)
    )(BaseComponent);

    return render(<Component />);
  };
}

afterEach(cleanup);

test('should throw an error when invalid argument', () => {
  expect(() => withMiddleware('counter', 'setState')).toThrowError();
});

describe('basic', () => {
  const renderer = createRenderer(Basic);

  test('should render without errors', () => {
    const { container } = renderer('counter', 'setCounter', 0, []);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should update state properly', () => {
    const { getByTestId } = renderer('counter', 'setCounter', 0, []);

    expect(getByTestId('value').textContent).toBe('0');

    fireEvent.click(getByTestId('increment'));
    expect(getByTestId('value').textContent).toBe('1');

    fireEvent.click(getByTestId('decrement'));
    expect(getByTestId('value').textContent).toBe('0');
  });
});

describe('using reducers', () => {
  const renderer = createRenderer(Dispatch);

  test('should render without errors', () => {
    const { container } = renderer('state', 'dispatch', reducer(), [
      useReducer(reducer)
    ]);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should update state properly', () => {
    const { getByTestId } = renderer('state', 'dispatch', reducer(), [
      useReducer(reducer)
    ]);

    expect(getByTestId('value').textContent).toBe('0');

    fireEvent.click(getByTestId('increment'));
    expect(getByTestId('value').textContent).toBe('1');

    fireEvent.click(getByTestId('decrement'));
    expect(getByTestId('value').textContent).toBe('0');
  });
});
