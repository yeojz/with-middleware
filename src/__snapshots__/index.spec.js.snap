// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic should render without errors 1`] = `
<div>
  <div
    data-testid="value"
  >
    0
  </div>
  <button
    data-testid="increment"
  >
    Increment
  </button>
  <button
    data-testid="decrement"
  >
    Decrement
  </button>
</div>
`;

exports[`using reducers should render without errors 1`] = `
<div>
  <div
    data-testid="value"
  >
    0
  </div>
  <button
    data-testid="increment"
  >
    Increment
  </button>
  <button
    data-testid="decrement"
  >
    Decrement
  </button>
</div>
`;
